### ISSUE TITLE
macOS: Can HttpClient do custom cert handling with libcurl+openssl10?

### ISSUE THREAD

**Issue by bartonjs** [dotnet/corefx#19718]
Gedankenexperiment:

* If OpenSSL/1.0 is the reported backend, run the OpenSSL initializer.
* Still fail client auth certificates (non-exportable private keys can't marshal into OpenSSL, and other problems could exist, so just block it)
* If OpenSSL/1.0 is the reported backend, register the callbacks/etc as normal.
* For the EE (server identity) cert and any presented intermediates, [extract the certificate bytes](https://github.com/dotnet/corefx/blob/5fe5f9aae7b2987adc7082f90712b265bee5eefc/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509CertificateReader.cs#L144-L147) to pass to `new X509Certificate2(byte[])`.
* [Build the X509Chain](https://github.com/dotnet/corefx/blob/0eeb35b0add4a8d1041a56de85409526c9466cac/src/System.Net.Security/src/System/Net/Security/SecureChannel.cs#L1012-L1022)
* Apply [chain and hostname checks](https://github.com/dotnet/corefx/blob/3b19899963606a34e1623e0926b90fe907197ee6/src/System.Net.Security/src/System/Net/CertificateValidationPal.OSX.cs#L13-L46)
* Call the callback.

**Comment by bartonjs** [comment]
cc @stephentoub

**Comment by karelz** [comment]
@bartonjs can you please set milestone? Or does it need discussion first?

**Comment by bartonjs** [comment]
Since this will enable a workaround on macOS where none currently exists, this seems pretty worthwhile for 2.0.  Of course, it's entirely experimental, so it's possible that it can't be done at a shipping quality in the time remaining.

**Comment by joeyaiello** [comment]
Just wanted to 👍 this with the context that it would unblock PowerShell/PowerShell#3648 which would help a non-trivial amount of our customers and partners use PowerShell Core 6.0 on macOS. Feel free to ping me offline if you need more formal justification.

**Comment by stephentoub** [comment]
@joeyaiello, just to be sure, https://github.com/dotnet/corefx/issues/19709 would also unblock PowerShell, right?  It looks like SkipCertificateCheck just uses `delegate { return true; }`, so once #19709 is implemented, PowerShell could just assign this delegate instead of that one, and it wouldn't require the end user to work around it by changing out the libcurl being used. (To be clear, I would still like to see this issue implemented, as it addresses a larger set of problems albeit with a more complicated workaround on the user's part.)

**Comment by joeyaiello** [comment]
@stephentoub that sounds right to me, but I would need @daxian-dbw to confirm. Going to linkback to #19709 in that issue as well. 

Thanks!

**Comment by karelz** [comment]
@bartonjs is the one you created workaround for? If not what else is left? Do you plan to finish it before vacation or should we give it to someone else?

**Comment by bartonjs** [comment]
@karelz I haven't had a chance to start yet, honestly.  If new problems would stop being discovered and distracting me, I'd probably be able to finish it.  But it's looking iffy.

**Comment by leecow** [comment]
Discussed in shiproom based on input from @stephentoub. Plan to work on for 2.1 and backport if it does turn out to be blocking.

**Comment by karelz** [comment]
@stephentoub did you mean to close it? Or leave it open for 2.1?

**Comment by karelz** [comment]
I think this was meant to track the 2.1 work ... reopening. @stephentoub please let me know if I misunderstood.

**Comment by tritao** [comment]
Any news when we can expect a fix for this one?

We're hitting this in really basic HTTP client code.

**Comment by karelz** [comment]
@bartonjs do you plan to work on it for 2.1?
cc @Priya91

**Comment by bartonjs** [comment]
I was given to understand that this would be rendered moot by the managed HTTP handler.  So I guess I thought this issue was closed.  Though apparently it's assigned to me, and open :smile:.

**Comment by karelz** [comment]
ManagedHandler will be just opt-in in 2.1. One day, it will be default.

Given the number of upvotes (22), I wonder if there is something targeted we could/should do for 2.1 -- given that we were even considering backporting it into servicing of 2.0 (https://github.com/dotnet/corefx/issues/19718#issuecomment-303532816).

**Comment by karelz** [comment]
We discussed it with @bartonjs couple of weeks ago.
We came to conclusion there is nothing reasonable we can do here, except to wait for SocketsHttpHandler, which is coming in 2.1 (maybe even as default).

Anyone can give it a try with SocketsHttpHandler in 2.1 Preview1 - see [blog post](https://blogs.msdn.microsoft.com/dotnet/2018/02/27/announcing-net-core-2-1-preview-1/) with details how to turn it on (heads up: the way how to turn it on will change in Preview2).
